---
title: Microbiome Analysis with DADA2 and Phyloseq
author: Ryan Johnson
date: '2020-01-29'
slug: microbiome-analysis-with-dada2-and-phyloseq
output:
  blogdown::html_page:
    toc: TRUE
categories:
  - R
  - dada2
  - phyloseq
  - microbiome
tags: []
image:
  caption: ''
  focal_point: ''
---


<div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#set-up-your-environment">Set Up Your Environment</a><ul>
<li><a href="#the-microbiome-data">The Microbiome Data</a></li>
<li><a href="#download-and-install-necessary-r-packages">Download and Install necessary R packages</a></li>
</ul></li>
<li><a href="#set-up-working-environment">Set up Working Environment</a><ul>
<li><a href="#reads">Reads</a></li>
</ul></li>
<li><a href="#the-analysis">The Analysis</a><ul>
<li><a href="#check-read-quality">Check Read Quality</a></li>
</ul></li>
<li><a href="#read-filtering">Read Filtering</a></li>
<li><a href="#learn-the-error-rates-and-infer-sequences">Learn the Error Rates and Infer Sequences</a></li>
</ul>
</div>

<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The purpose of this post will be to guide researchers through a basic analysis of microbiome data using R packages <code>DADA2</code> and <code>Phyloseq</code>. Most concepts will be discussed at a very high level and I won’t spend too much time digging into the weeds of the analysis.</p>
<p>We will be analyzing a very small subset of data that was used in part to look at differences in microbiome structure between mice given a regular diet (<strong>RD</strong>, n = 24) versus a diet with no isoflavones (<strong>NIF</strong>, n = 24). Fecal samples from each mouse were collected 2 weeks after being fed either the RD or NIF diets. Samples were processed in the lab and subjected to Illumina MiSeq 300 base paired-end sequencing. We specifically targeted the V4 variable region of the 16S rRNA gene for sequencing. Reads from each sample were subsampled to 5000 reads/sample just to make the data a bit more manageable.</p>
</div>
<div id="set-up-your-environment" class="section level1">
<h1>Set Up Your Environment</h1>
<p>Before we can get started, there are a few things you’ll need to download/install:</p>
<div id="the-microbiome-data" class="section level3">
<h3>The Microbiome Data</h3>
<p>The data has been compressed into a single <code>tar.gz</code> file. You’ll need to download it and de-compress it. This can usually be done by simply double-clicking on it.</p>
<p>Download the data <a href="https://drive.google.com/file/d/1HCcMQASYweEYrjTQUxuHeaQ07FlJf-3U/view?usp=sharing">here</a></p>
</div>
<div id="download-and-install-necessary-r-packages" class="section level3">
<h3>Download and Install necessary R packages</h3>
<p>In order to get <code>DADA2</code>, <code>Phyloseq</code>, and a few other packages installed on your computer, you need to install them from the internet. Some of these packages can take a while to install, so don’t be alarmed if it take a couple minutes. <code>DADA2</code> and <code>Phyloseq</code> are held within Bioconductor, which a collection of packages used primarily for biological data analysis, so you’ll need to install Bioconductor prior to installing <code>DADA2</code> and <code>Phyloseq</code>. Once all packags are installed, you won’t have access to them until you “turn them on” using the <code>library()</code> command.</p>
<ul>
<li><p><span style="color: red;"><strong>DADA2</strong></span>: Please follow the directions from this <a href="https://benjjneb.github.io/dada2/dada-installation.html">website</a>. This is the package that does much of the “heavy” lifting in terms of read quality processing and error rate detection.</p></li>
<li><p><span style="color: red;"><strong>Phyloseq</strong></span>: Please follow the directions from this <a href="https://bioconductor.org/packages/release/bioc/html/phyloseq.html">website</a>. This package is primarily used to process the high-quality reads, generate diversity statistics, and make pretty figures.</p></li>
<li><p><span style="color: red;"><strong>Tidyverse</strong></span>: This is a suite of packages that are used for variety of data science needs (read/write files, clean/edit data, make cool figures, etc). To install, simply enter <code>install.packages(&quot;tidyverse&quot;)</code> into your R prompt.</p></li>
</ul>
<p>These two blocks of code should install and load all of the packages needed:</p>
<pre class="r"><code># DADA2 install
if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;dada2&quot;, version = &quot;3.10&quot;)

# Phyloseq install
if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;phyloseq&quot;)

# Tidyverse install
install.packages(&quot;tidyverse&quot;)</code></pre>
<pre class="r"><code># Load packages
library(dada2)
library(phyloseq)
library(tidyverse)</code></pre>
</div>
</div>
<div id="set-up-working-environment" class="section level1">
<h1>Set up Working Environment</h1>
<div id="reads" class="section level3">
<h3>Reads</h3>
<p>To make things easier, we need to create some variables that will hold the names of our fastq files and the directory for our soon-to-be processed reads. The following code block will store the paths for the forward (R1) and reverse (R2) fastq files (raw and processed):</p>
<pre class="r"><code># Get path of directory that contains all of the reads
#  change if reads in separate file on your computer
reads_path &lt;- &quot;data/mouse_samples&quot; 

# Store path of forward and reverse reads
Fs_path &lt;- sort(list.files(reads_path, pattern=&quot;R1.fastq&quot;, full.names = TRUE))
Rs_path &lt;- sort(list.files(reads_path, pattern=&quot;R2.fastq&quot;, full.names = TRUE))

# Create directories for processed forward and reverse reads
Fs_path_filtered &lt;- file.path(reads_path, &quot;filtered_Fs&quot;)
Rs_path_filtered &lt;- file.path(reads_path, &quot;filtered_Rs&quot;)</code></pre>
<p>It will also be useful later on to store the sample names without the R1/R2 and fastq annotation:</p>
<pre class="r"><code>mouse_sample_names &lt;- str_replace(string = basename(Fs_path), 
                                  pattern = &quot;_R1\\.fastq&quot;,
                                  replacement = &quot;&quot;)</code></pre>
</div>
</div>
<div id="the-analysis" class="section level1">
<h1>The Analysis</h1>
<div id="check-read-quality" class="section level3">
<h3>Check Read Quality</h3>
<p>We first want to get a general idea of the quality of the reads in our dataset. Let’s look at a random subsampling of the samples:</p>
<pre class="r"><code>set.seed(1234) # Ensures the same &quot;random&quot; sampling is performed

# Forward Read quality
plotQualityProfile(Fs_path[sample(1:48, 12, replace = FALSE)])</code></pre>
<p><img src="/post/2020-01-29-microbiome-analysis-with-dada2-and-phyloseq_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<pre class="r"><code># Reverse Read quality
plotQualityProfile(Rs_path[sample(1:48, 12, replace = FALSE)])</code></pre>
<p><img src="/post/2020-01-29-microbiome-analysis-with-dada2-and-phyloseq_files/figure-html/unnamed-chunk-3-2.png" width="672" /></p>
<p>We typically don’t want to have the read quality drop below ~30 (focus on the green lines which represent mean quality). But we also need to make sure that our reads have sufficient overlap. The V4 region of the 16S rRNA gene is about 250bp in length, so if we decide to trim off the ends of the sequences due to low quality, the remaining lengths must be &gt;250…actually it needs to be more than that in order to have sufficient overlap. I believe that DADA2 requires at least 12bp overlap, but the more the better.</p>
<p>For the forward reads, we see the quality drop at around <strong>200bp</strong>. For the reverse reads, which are lower in quality (not atypical), we see the quality drop at around <strong>150bp</strong>. We will use 200 and 150 to trim our seqeunces.</p>
</div>
</div>
<div id="read-filtering" class="section level1">
<h1>Read Filtering</h1>
<p>Using the following parameters:</p>
<ul>
<li><p><code>maxN=0</code> (DADA2 requires no Ns)</p></li>
<li><p><code>truncQ=2</code>Truncate reads at the first instance of a quality score less than or equal to truncQ (keeping this as default)</p></li>
<li><p><code>rm.phix=TRUE</code>: discard reads that match against the phiX genome</p></li>
<li><p><code>maxEE=c(2,2)</code>: sets the maximum number of “expected errors” allowed in a read, which is a better filter than simply averaging quality scores.</p></li>
</ul>
<pre class="r"><code>out &lt;- filterAndTrim(fwd=Fs_path, 
              filt=Fs_path_filtered,
              rev=Rs_path, 
              filt.rev=Rs_path_filtered,
              truncLen=c(200,150), # forward and reverse read 
              maxEE=c(2,2), 
              truncQ=2, 
              maxN=0, 
              rm.phix=TRUE,
              compress=TRUE, 
              verbose=TRUE, 
              multithread=TRUE)
out</code></pre>
<pre><code>##                reads.in reads.out
## NIF1_R1.fastq      5000      4013
## NIF10_R1.fastq     5000      4137
## NIF11_R1.fastq     5000      4320
## NIF12_R1.fastq     5000      4075
## NIF13_R1.fastq     5000      4111
## NIF14_R1.fastq     5000      4370
## NIF15_R1.fastq     5000      4075
## NIF16_R1.fastq     5000      4317
## NIF17_R1.fastq     5000      4169
## NIF18_R1.fastq     5000      4108
## NIF19_R1.fastq     5000      4091
## NIF2_R1.fastq      5000      4121
## NIF20_R1.fastq     5000      4121
## NIF21_R1.fastq     5000      4061
## NIF22_R1.fastq     5000      4542
## NIF23_R1.fastq     5000      4203
## NIF24_R1.fastq     5000      4219
## NIF3_R1.fastq      5000      4358
## NIF4_R1.fastq      5000      4247
## NIF5_R1.fastq      5000      4260
## NIF6_R1.fastq      5000      4314
## NIF7_R1.fastq      5000      4066
## NIF8_R1.fastq      5000      4056
## NIF9_R1.fastq      5000      4071
## RD1_R1.fastq       5000      4173
## RD10_R1.fastq      5000      4264
## RD11_R1.fastq      5000      4276
## RD12_R1.fastq      5000      4208
## RD13_R1.fastq      5000      4272
## RD14_R1.fastq      5000      4189
## RD15_R1.fastq      5000      4390
## RD16_R1.fastq      5000      4182
## RD17_R1.fastq      5000      4529
## RD18_R1.fastq      5000      4135
## RD19_R1.fastq      5000      4090
## RD2_R1.fastq       5000      4103
## RD20_R1.fastq      5000      4253
## RD21_R1.fastq      5000      4214
## RD22_R1.fastq      5000      4357
## RD23_R1.fastq      5000      4539
## RD24_R1.fastq      5000      4212
## RD3_R1.fastq       5000      4230
## RD4_R1.fastq       5000      4341
## RD5_R1.fastq       5000      4233
## RD6_R1.fastq       5000      4231
## RD7_R1.fastq       5000      4121
## RD8_R1.fastq       5000      4244
## RD9_R1.fastq       5000      4272</code></pre>
<p>We can see from the output that for most samples, about 600-900 reads were removed from each sample. This seems pretty reasonable. If you find that most of your reads are being thrown out, then you may need to tweak your filtering parameters.</p>
</div>
<div id="learn-the-error-rates-and-infer-sequences" class="section level1">
<h1>Learn the Error Rates and Infer Sequences</h1>
<p>High throughput sequencing is not perfect…in fact is pretty error prone. Sometimes you’ll see workflows estimate their error rates by sequencing a mock community (which is a good idea). However, DADA2 leverages some statistical magic (a parametric error model) to estimate sequencing errors by comparing amplicons to the most abundant amplicons present in your sample. The default is to use 1x10<sup>8</sup> (recommended), but to spead things up, I’ll use 1e<sup>7</sup> reads.</p>
<pre class="r"><code># Forward read estimates
errF &lt;- learnErrors(Fs_path_filtered, nbases = 1e7, multithread=TRUE)</code></pre>
<pre><code>## 10805600 total bases in 54028 reads from 13 samples will be used for learning the error rates.</code></pre>
<pre class="r"><code># Reverse read estimates
errF &lt;- learnErrors(Rs_path_filtered, nbases = 1e7, multithread=TRUE)</code></pre>
<pre><code>## 10025100 total bases in 66834 reads from 16 samples will be used for learning the error rates.</code></pre>
<p>Now that you’ve estimated the error rates, we need to go back to our samples and analyze each read and infer it’s true sequence given the error rates. Basically, it’s going to determine the number of <em>unique</em> reads per sample.</p>
<pre class="r"><code># Infer forward reads
dadaFs &lt;- dada(Fs_path_filtered, err=errF, multithread=TRUE)</code></pre>
<pre><code>## Sample 1 - 4013 reads in 1776 unique sequences.
## Sample 2 - 4137 reads in 1641 unique sequences.
## Sample 3 - 4320 reads in 1653 unique sequences.
## Sample 4 - 4075 reads in 1717 unique sequences.
## Sample 5 - 4111 reads in 1770 unique sequences.
## Sample 6 - 4370 reads in 1761 unique sequences.
## Sample 7 - 4075 reads in 1574 unique sequences.
## Sample 8 - 4317 reads in 1929 unique sequences.
## Sample 9 - 4169 reads in 1876 unique sequences.
## Sample 10 - 4108 reads in 1922 unique sequences.
## Sample 11 - 4091 reads in 1811 unique sequences.
## Sample 12 - 4121 reads in 1744 unique sequences.
## Sample 13 - 4121 reads in 1893 unique sequences.
## Sample 14 - 4061 reads in 1823 unique sequences.
## Sample 15 - 4542 reads in 1915 unique sequences.
## Sample 16 - 4203 reads in 1773 unique sequences.
## Sample 17 - 4219 reads in 1835 unique sequences.
## Sample 18 - 4358 reads in 1580 unique sequences.
## Sample 19 - 4247 reads in 1693 unique sequences.
## Sample 20 - 4260 reads in 1795 unique sequences.
## Sample 21 - 4314 reads in 1735 unique sequences.
## Sample 22 - 4066 reads in 1598 unique sequences.
## Sample 23 - 4056 reads in 1667 unique sequences.
## Sample 24 - 4071 reads in 1624 unique sequences.
## Sample 25 - 4173 reads in 1854 unique sequences.
## Sample 26 - 4264 reads in 1794 unique sequences.
## Sample 27 - 4276 reads in 2022 unique sequences.
## Sample 28 - 4208 reads in 1871 unique sequences.
## Sample 29 - 4272 reads in 1911 unique sequences.
## Sample 30 - 4189 reads in 1834 unique sequences.
## Sample 31 - 4390 reads in 1768 unique sequences.
## Sample 32 - 4182 reads in 1851 unique sequences.
## Sample 33 - 4529 reads in 1839 unique sequences.
## Sample 34 - 4135 reads in 1782 unique sequences.
## Sample 35 - 4090 reads in 1773 unique sequences.
## Sample 36 - 4103 reads in 1937 unique sequences.
## Sample 37 - 4253 reads in 1758 unique sequences.
## Sample 38 - 4214 reads in 1781 unique sequences.
## Sample 39 - 4357 reads in 2018 unique sequences.
## Sample 40 - 4539 reads in 1643 unique sequences.
## Sample 41 - 4212 reads in 1679 unique sequences.
## Sample 42 - 4230 reads in 2000 unique sequences.
## Sample 43 - 4341 reads in 1931 unique sequences.
## Sample 44 - 4233 reads in 1798 unique sequences.
## Sample 45 - 4231 reads in 1769 unique sequences.
## Sample 46 - 4121 reads in 1796 unique sequences.
## Sample 47 - 4244 reads in 1887 unique sequences.
## Sample 48 - 4272 reads in 1702 unique sequences.</code></pre>
<pre class="r"><code># Infer reverse reads
dadaRs &lt;- dada(Rs_path_filtered, err=errF, multithread=TRUE)</code></pre>
<pre><code>## Sample 1 - 4013 reads in 1762 unique sequences.
## Sample 2 - 4137 reads in 1692 unique sequences.
## Sample 3 - 4320 reads in 1713 unique sequences.
## Sample 4 - 4075 reads in 1679 unique sequences.
## Sample 5 - 4111 reads in 1770 unique sequences.
## Sample 6 - 4370 reads in 1713 unique sequences.
## Sample 7 - 4075 reads in 1634 unique sequences.
## Sample 8 - 4317 reads in 1883 unique sequences.
## Sample 9 - 4169 reads in 1878 unique sequences.
## Sample 10 - 4108 reads in 1891 unique sequences.
## Sample 11 - 4091 reads in 1747 unique sequences.
## Sample 12 - 4121 reads in 1728 unique sequences.
## Sample 13 - 4121 reads in 1975 unique sequences.
## Sample 14 - 4061 reads in 1790 unique sequences.
## Sample 15 - 4542 reads in 1991 unique sequences.
## Sample 16 - 4203 reads in 1850 unique sequences.
## Sample 17 - 4219 reads in 1796 unique sequences.
## Sample 18 - 4358 reads in 1625 unique sequences.
## Sample 19 - 4247 reads in 1727 unique sequences.
## Sample 20 - 4260 reads in 1832 unique sequences.
## Sample 21 - 4314 reads in 1697 unique sequences.
## Sample 22 - 4066 reads in 1617 unique sequences.
## Sample 23 - 4056 reads in 1679 unique sequences.
## Sample 24 - 4071 reads in 1669 unique sequences.
## Sample 25 - 4173 reads in 1850 unique sequences.
## Sample 26 - 4264 reads in 1697 unique sequences.
## Sample 27 - 4276 reads in 1918 unique sequences.
## Sample 28 - 4208 reads in 1805 unique sequences.
## Sample 29 - 4272 reads in 1892 unique sequences.
## Sample 30 - 4189 reads in 1810 unique sequences.
## Sample 31 - 4390 reads in 1750 unique sequences.
## Sample 32 - 4182 reads in 1905 unique sequences.
## Sample 33 - 4529 reads in 1790 unique sequences.
## Sample 34 - 4135 reads in 1749 unique sequences.
## Sample 35 - 4090 reads in 1772 unique sequences.
## Sample 36 - 4103 reads in 1911 unique sequences.
## Sample 37 - 4253 reads in 1720 unique sequences.
## Sample 38 - 4214 reads in 1720 unique sequences.
## Sample 39 - 4357 reads in 2002 unique sequences.
## Sample 40 - 4539 reads in 1614 unique sequences.
## Sample 41 - 4212 reads in 1695 unique sequences.
## Sample 42 - 4230 reads in 1963 unique sequences.
## Sample 43 - 4341 reads in 1941 unique sequences.
## Sample 44 - 4233 reads in 1744 unique sequences.
## Sample 45 - 4231 reads in 1668 unique sequences.
## Sample 46 - 4121 reads in 1800 unique sequences.
## Sample 47 - 4244 reads in 1831 unique sequences.
## Sample 48 - 4272 reads in 1654 unique sequences.</code></pre>
</div>
