---
title: Microbiome Analysis with DADA2 and Phyloseq
author: Ryan Johnson
date: '2020-01-29'
slug: microbiome-analysis-with-dada2-and-phyloseq
output:
  blogdown::html_page:
    toc: TRUE
categories:
  - R
  - dada2
  - phyloseq
  - microbiome
tags: []
image:
  caption: ''
  focal_point: ''
---

# Introduction

The purpose of this post will be to guide researchers through a basic analysis of microbiome data using R packages `DADA2` and `Phyloseq`. Most concepts will be discussed at a very high level and I won't spend too much time digging into the weeds of the analysis. 

We will be analyzing a very small subset of data that was used in part to look at differences in microbiome structure between mice given a regular diet (**RD**, n = 24) versus a diet with no isoflavones (**NIF**, n = 24). Fecal samples from each mouse were collected 2 weeks after being fed either the RD or NIF diets. Samples were processed in the lab and subjected to Illumina MiSeq 300 base paired-end sequencing. We specifically targeted the V4 variable region of the 16S rRNA gene for sequencing. Reads from each sample were subsampled to 5000 reads/sample just to make the data a bit more manageable.

# Set Up Your Environment

Before we can get started, there are a few things you'll need to download/install:

### The Microbiome Data

The data has been compressed into a single `tar.gz` file. You'll need to download it and de-compress it. This can usually be done by simply double-clicking on it.

Download the data [here](https://drive.google.com/file/d/1HCcMQASYweEYrjTQUxuHeaQ07FlJf-3U/view?usp=sharing)

### Download and Install necessary R packages

In order to get `DADA2`, `Phyloseq`, and a few other packages installed on your computer, you need to install them from the internet. Some of these packages can take a while to install, so don't be alarmed if it take a couple minutes. `DADA2` and `Phyloseq` are held within Bioconductor, which a collection of packages used primarily for biological data analysis, so you'll need to install Bioconductor prior to installing `DADA2` and `Phyloseq`. Once all packags are installed, you won't have access to them until you "turn them on" using the `library()` command.

* <span style="color: red;">**DADA2**</span>: Please follow the directions from this [website](https://benjjneb.github.io/dada2/dada-installation.html). This is the package that does much of the "heavy" lifting in terms of read quality processing and error rate detection.

* <span style="color: red;">**Phyloseq**</span>: Please follow the directions from this [website](https://bioconductor.org/packages/release/bioc/html/phyloseq.html). This package is primarily used to process the high-quality reads, generate diversity statistics, and make pretty figures.

* <span style="color: red;">**Tidyverse**</span>: This is a suite of packages that are used for variety of data science needs (read/write files, clean/edit data, make cool figures, etc). To install, simply enter `install.packages("tidyverse")` into your R prompt.

These two blocks of code should install and load all of the packages needed:

```{r install_packages, eval=FALSE, message=FALSE, warning=FALSE}
# DADA2 install
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("dada2", version = "3.10")

# Phyloseq install
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("phyloseq")

# Tidyverse install
install.packages("tidyverse")
```


```{r load_libraries, warning=FALSE, message=FALSE}
# Load packages
library(dada2)
library(phyloseq)
library(tidyverse)
```



# Set up Working Environment

### Reads

To make things easier, we need to create some variables that will hold the names of our fastq files and the directory for our soon-to-be processed reads. The following code block will store the paths for the forward (R1) and reverse (R2) fastq files (raw and processed):

```{r}
# Get path of directory that contains all of the reads
#  change if reads in separate file on your computer
reads_path <- "data/mouse_samples" 

# Store path of forward and reverse reads
Fs_path <- sort(list.files(reads_path, pattern="R1.fastq", full.names = TRUE))
Rs_path <- sort(list.files(reads_path, pattern="R2.fastq", full.names = TRUE))

# Create directories for processed forward and reverse reads
Fs_path_filtered <- file.path(reads_path, "filtered_Fs")
Rs_path_filtered <- file.path(reads_path, "filtered_Rs")
```


It will also be useful later on to store the sample names without the R1/R2 and fastq annotation:

```{r}
mouse_sample_names <- str_replace(string = basename(Fs_path), 
                                  pattern = "_R1\\.fastq",
                                  replacement = "")
```



# The Analysis

### Check Read Quality

We first want to get a general idea of the quality of the reads in our dataset. Let's look at a random subsampling of the samples:

```{r cache=TRUE}
set.seed(1234) # Ensures the same "random" sampling is performed

# Forward Read quality
plotQualityProfile(Fs_path[sample(1:48, 12, replace = FALSE)])

# Reverse Read quality
plotQualityProfile(Rs_path[sample(1:48, 12, replace = FALSE)])
```

We typically don't want to have the read quality drop below ~30 (focus on the green lines which represent mean quality). But we also need to make sure that our reads have sufficient overlap. The V4 region of the 16S rRNA gene is about 250bp in length, so if we decide to trim off the ends of the sequences due to low quality, the remaining lengths must be >250...actually it needs to be more than that in order to have sufficient overlap. I believe that DADA2 requires at least 12bp overlap, but the more the better.

For the forward reads, we see the quality drop at around **200bp**. For the reverse reads, which are lower in quality (not atypical), we see the quality drop at around **150bp**. We will use 200 and 150 to trim our seqeunces.


# Read Filtering

Using the following parameters:

* `maxN=0` (DADA2 requires no Ns)

* `truncQ=2`Truncate reads at the first instance of a quality score less than or equal to truncQ (keeping this as default)

* `rm.phix=TRUE`: discard reads that match against the phiX genome

* `maxEE=c(2,2)`: sets the maximum number of “expected errors” allowed in a read, which is a better filter than simply averaging quality scores.

```{r cache=TRUE}
out <- filterAndTrim(fwd=Fs_path, 
              filt=Fs_path_filtered,
              rev=Rs_path, 
              filt.rev=Rs_path_filtered,
              truncLen=c(200,150), # forward and reverse read 
              maxEE=c(2,2), 
              truncQ=2, 
              maxN=0, 
              rm.phix=TRUE,
              compress=TRUE, 
              verbose=TRUE, 
              multithread=TRUE)
out
```

We can see from the output that for most samples, about 600-900 reads were removed from each sample.